src/lib/validation/validators/required.ts

```diff
@@ -0,0 +1,24 @@
+import { ValidationError } from '@lib/validation/validationerror'
+import { IComponentProps } from '@types'
+
+/**
+ * Validates whether value is truthy when component is required.
+ * @throws {RequiredValidationError} As promise rejection if invalid.
+ */
```

Comment:
to cut this discussion short - check the reference behaviour of formio. what happens in their formbuilder if you specify required AND a pattern, if you get both validation errors -> we also do both validation errors (implementation wise that's the easiest). It may not be the ideal user experience, but let's settle for "not worse than the current implementation" at first and iterate from there.

Other than that, w/r to pristine: I'd say it's not pristine as soon as the value is different from the component defaultValue. This implies that it was touched and changed by the user and not just that they had focus and blur on the input.

Otherwise, an empty value does depend on the component type: textfield is empty for null, undefined and empty string, while checkbox is only empty for `null` value. Multiple checkboxes empty -> I'm not sure, check the formio behaviour for reference.

The variant of `component.multiple` is also one to take into account, as you now have an Array type and empty means `Array.length === 0`